src/development/metrics.md

```diff
@@ -0,0 +1,153 @@
+
+![image.png](./assets/hash_speed_diff.png)
+
+At the end of the sha256 run, so >200k blocks later, the chain was 75072 blocks ahead. This represents a loss of 3 hours per day against sha256. That’s >12% faster.
+
+Coupled to the fact that Pedersen hashes aren’t quantum resistant, it seems like a good idea to reconsider using Pedersen hashes to compute trees.
+
+## Verkle-specific measurements
+
+### Context
+
+This data is gathered replaying ~200k historical blocks around the Shanghai fork. While they do not give a perfect view of how execution will behave once EIP-4762 is active, it still gives a good indication of what to expect, and how the spec should evolve.
+
+### Witnesses
+
+#### Size
+
+Research has been conducted, in order to determine the best structure for a witness. We tried three methods:
+
+ * The first one, as specified in the [current version of the consensus spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip6800/beacon-chain.md#executionwitness), referred to as **type 0**. This approach uses the SSZ type `Optional[T]` , that is yet to be officially included in the spec.
+ * A first variation of the spec, in which prestate values are grouped together in their own list, and poststate values are grouped in their own list as well. This is referred to as **type 1**.
+
+```python
+class SuffixStateDiffs(Container):
+    suffixes: bytes
+    # None means not currently present
+    current_values: List[bytes]
+    # None means value not updated
+    new_values:     List[bytes]
+```
+Type 1 serialization container
+
+ * A second variation of the spec, in which updates, insertions and reads are grouped in their own separate lists. The suffixes for each of these lists are also grouped as their own byte lists. This is referred to as **type 2**.
+
+```python
+class SuffixStateDiffs(Container):
+    updated_suffixes: bytes
+    updated_currents: List[Bytes32]
+    updated_news: List[Bytes32]
+
+    inserted_suffixes: bytes
+    inserted_news: List[Bytes32]
+
+    read_suffixes: bytes
+    read_currents: List[Bytes32]
+
+    missing_suffixes: bytes
+```
+Type 2 serialization container
+
+ * A final variation, where update, reads, inserts and proof-of-absences are grouped, along with their suffixes, into their own lists. This is referred to as **type 3**.
+
+```python
+class UpdateDiff(Container):
+    suffix: byte
+    current: Bytes32
+    new: Bytes32
+
+class ReadDiff(Container):
+    suffix: byte
+    current: Bytes32
+
+class InsertDiff(Container):
+    suffix: byte
+    new: Bytes32
+
+class MissingDiff(Container):
+    suffix byte
+
+class StemStateDiff(Container):
+    stem: Bytes31
+
+    updates: List[UpdateDiff]
+    reads: List[ReadStateDiff]
+    insert: List[InsertStateDiff]
+    # TODO: check if this encodes as well as bytes
+    missing: List[MissingStateDiff]
+```
+Type 3 serialization container
+
+Here are our findings after replaying data for 200k blocks:
+
+![image.png](./assets/total.png)
```

Comment:
What does `during/post-transition` means?
Does this mean that the witnesses are partial since the tree conversion is running?

Reply:
yeah, a lot of the `CurrentValue`s are `null`

Reply:
they are useful for sync, potentially. So that's why I wanted to see what the impact on the witness during the transition was.